plugin/web/extensions/sidebar/popups/buildingSearch/components/hooks.ts

```diff
@@ -0,0 +1,214 @@
+  const [showMatchingOnly, setShowMatchingOnly] = useState<boolean>(false);
+  const [selected, setSelected] = useState<string[]>([]);
+
+  const conditionApply = useCallback(() => {
+    // TODO: Search logic here
+    console.log(conditions);
+    const computeResult = TEST_RESULT_DATA;
+
+    setResults(computeResult);
+    setActiveTab("result");
+    setHighlightAll(true);
+    setShowMatchingOnly(false);
+  }, [conditions]);
+
+  useEffect(() => {
+    // TODO: flyTo the selected building
+    if (selected.length === 1) {
+      postMsg({
+        action: "cameraFlyTo",
+        payload: [
+          {
+            lng: 137.31210397018728,
+            lat: 34.60832876429294,
+            height: 54735.3396536646,
+            heading: 0.06105070089725917,
+            pitch: -0.8254743840751195,
+            roll: 6.283184517370357,
+            fov: 1.0471975511965976,
+          },
+          { duration: 2 },
+        ],
+      });
+    }
+  }, [selected]);
+
+  useEffect(() => {
+    // TODO: flyTo the result if only one
+    if (results.length === 1) {
+      postMsg({
+        action: "cameraFlyTo",
+        payload: [
+          {
+            lng: 137.31210397018728,
+            lat: 34.60832876429294,
+            height: 54735.3396536646,
+            heading: 0.06105070089725917,
+            pitch: -0.8254743840751195,
+            roll: 6.283184517370357,
+            fov: 1.0471975511965976,
+          },
+          { duration: 2 },
+        ],
+      });
+    }
+  }, [results]);
+
+  useEffect(() => {
+    // TODO: Update 3D tiles style
+  }, [highlightAll, showMatchingOnly, selected, results]);
+
+  const popupClose = useCallback(() => {
+    postMsg({ action: "popupClose" });
+  }, []);
+
+  useEffect(() => {
+    if ((window as any).buildingSearchInit) {
+      const init = (window as any).buildingSearchInit;
+      if (init.viewport.isMobile) {
+        setIsMobile(true);
+        setSizes({
+          ...sizes,
+          mobile: { width: init.viewport.width * 0.9, height: sizes.mobile.height },
+        });
+      }
+    }
+
+    document.documentElement.style.setProperty("--theme-color", "#00BEBE");
+
+    const datasetIndexes = ((window as any).buildingSearchInit.data ??
+      TEST_DATASET_INDEX_DATA) as DatasetIndexes;
+
+    setDatasetIndexes(datasetIndexes);
+    setConditions(datasetIndexes.indexes.map(index => ({ field: index.field, values: [] })));
+
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const onMessage = useCallback(
+    (e: MessageEvent<any>) => {
+      if (e.source !== parent) return;
+      switch (e.data.type) {
+        case "resize":
+          handleResize(e.data.payload);
+          break;
+        default:
+          break;
+      }
+    },
+    [handleResize],
+  );
+
+  useEffect(() => {
+    addEventListener("message", onMessage);
+    return () => {
+      removeEventListener("message", onMessage);
+    };
+  }, [onMessage]);
```

Comment:
I think if you add anything to the deps array of a useEffect with addEventListener you will get multiple even listeners in the lifetime of the the component. Just move the onMessage function inside the useEffect.

Reply:
Emmm i think even move it inside, the `handleResize` will also become the dependency. I think it's reasonalble to list as dependency and when dependency changes the returned clear function will remove previous event listner so there won't be multiple, right?